nav2_collision_monitor/src/polygon_source.cpp

```diff
@@ -0,0 +1,215 @@
+    int num_points_in_segment =
+      std::max(static_cast<int>(std::ceil(segment_length / sampling_distance_)), 1);
+
+    // Calculate the step size for each pair of vertices
+    const double dx = (next_point.x - current_point.x) / num_points_in_segment;
+    const double dy = (next_point.y - current_point.y) / num_points_in_segment;
+
+    // Sample the points with equal spacing
+    for (int j = 0; j <= num_points_in_segment; ++j) {
+      Point p;
+      p.x = current_point.x + j * dx;
+      p.y = current_point.y + j * dy;
+      data.push_back(p);
+    }
+  }
+}
+
+void PolygonSource::getParameters(std::string & source_topic)
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  getCommonParameters(source_topic);
+
+  nav2_util::declare_parameter_if_not_declared(
+    node, source_name_ + ".sampling_distance", rclcpp::ParameterValue(0.1));
+  sampling_distance_ = node->get_parameter(source_name_ + ".sampling_distance").as_double();
+  nav2_util::declare_parameter_if_not_declared(
+    node, source_name_ + ".polygon_similarity_threshold", rclcpp::ParameterValue(2.0));
+  polygon_similarity_threshold_ =
+    node->get_parameter(source_name_ + ".polygon_similarity_threshold").as_double();
+}
+
+void PolygonSource::dataCallback(geometry_msgs::msg::PolygonStamped::ConstSharedPtr msg)
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+  auto curr_time = node->now();
+
+  // remove old data
+  data_.erase(
+    std::remove_if(
+      data_.begin(), data_.end(),
+      [this, curr_time](const geometry_msgs::msg::PolygonStamped & polygon_stamped) {
+        return curr_time - rclcpp::Time(polygon_stamped.header.stamp) > source_timeout_;
+      }), data_.end());
+
+  // check if older similar polygon exists already and replace it with the new one
+  for (auto & polygon_stamped : data_) {
+    if (msg->header.frame_id == polygon_stamped.header.frame_id &&
+      rclcpp::Time(msg->header.stamp) > rclcpp::Time(polygon_stamped.header.stamp) &&
+      arePolygonsSimilar(msg->polygon, polygon_stamped.polygon))
+    {
+      polygon_stamped = *msg;
+      return;
+    }
+  }
+  data_.push_back(*msg);
+}
+
+bool PolygonSource::arePolygonsSimilar(
+  const geometry_msgs::msg::Polygon & polygon1,
+  const geometry_msgs::msg::Polygon & polygon2) const
+{
+  /* Generated by GPT */
+
+  // Compare the number of points in the polygons
+  if (polygon1.points.size() != polygon2.points.size()) {
+    return false;
+  }
+
+  // Compare the distance between each pair of corresponding points
+  for (size_t i = 0; i < polygon1.points.size(); ++i) {
+    double dx = polygon1.points[i].x - polygon2.points[i].x;
+    double dy = polygon1.points[i].y - polygon2.points[i].y;
+    double dz = polygon1.points[i].z - polygon2.points[i].z;
+    double distance = std::sqrt(dx * dx + dy * dy + dz * dz);
```

Comment:
How does this generalize?

Reply:
I think if you want to replace data, we need to have (1) some kind of unique identifier or (2) have unique frames of each thing that we use TF to transform. (2) feels fragile, so I think the answer here is to have some unique identifier in the messages of the physical object the polygon belongs to. This feels like a good add to the standard messages too, let me inquire if we can make an API change

https://github.com/ros2/common_interfaces/issues/230

Reply:
For the unique ID, we could abuse the 'seq' of the Header msg. I say "abuse" because it's more meant as a sequence counter.

Regarding your PR, you could also make the case to adding the unique identifier to the Header msg, since this uid could be needed for many other msgs, e.g. Point, Vector, etc...

(I'm on vacation for the whole week; I'll try to answer but it might take a while or not be as detailled as I would wish)

Reply:
There is no seq in ROS 2 :smiling_face_with_tear: 

No worries, enjoy vacation. I'd be disappointed if I heard from you the rest of the week :P

Reply:
Damn it, you're right I was looking at the ROS 1 version. Second part of my comment is still valid though.

Reply:
While I agree that an id in the header could be useful, I think that is a uphill battle, but is worth a discussion!